File: apps/flutter_parent/lib/models/submission.dart
// Copyright (C) 2019 - present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
library submission;

import 'package:built_collection/built_collection.dart';
import 'package:built_value/built_value.dart';
import 'package:built_value/serializer.dart';

import 'assignment.dart';
import 'user.dart';

part 'submission.g.dart';

abstract class Submission implements Built<Submission, SubmissionBuilder> {
  @BuiltValueSerializer(serializeNulls: true)
  static Serializer<Submission> get serializer => _$submissionSerializer;

  Submission._();

  factory Submission([void Function(SubmissionBuilder) updates]) = _$Submission;

  String get id;

  String? get grade;

  double get score;

  int get attempt;

  @BuiltValueField(wireName: 'submitted_at')
  DateTime? get submittedAt;

  DateTime? get commentCreated;

  String? get mediaContentType;

  String? get mediaCommentUrl;

  String? get mediaCommentDisplay;

  @BuiltValueField(wireName: 'submission_history')
  BuiltList<Submission> get submissionHistory;

  String? get body;

  @BuiltValueField(wireName: 'grade_matches_current_submission')
  bool get isGradeMatchesCurrentSubmission;

  @BuiltValueField(wireName: 'workflow_state')
  String? get workflowState;

  @BuiltValueField(wireName: 'submission_type')
  String? get submissionType;

  @BuiltValueField(wireName: 'preview_url')
  String? get previewUrl;

  String? get url;

  // Not sure why, but build_runner fails when this field is named 'late'
  @BuiltValueField(wireName: 'late')
  bool get isLate;

  bool get excused;

  bool get missing;

  // Conversation Stuff
  @BuiltValueField(wireName: 'assignment_id')
  String get assignmentId;

  Assignment? get assignment;

  @BuiltValueField(wireName: 'user_id')
  String get userId;

  @BuiltValueField(wireName: 'grader_id')
  String get graderId;

  User? get user;

  @BuiltValueField(wireName: 'points_deducted')
  double? get pointsDeducted;

  @BuiltValueField(wireName: 'entered_score')
  double get enteredScore;

  @BuiltValueField(wireName: 'entered_grade')
  String? get enteredGrade;

  @BuiltValueField(wireName: 'posted_at')
  DateTime? get postedAt;

  bool isGraded() {
    return grade != null && workflowState != 'pending_review' && postedAt != null;
  }

  static void _initializeBuilder(SubmissionBuilder b) => b
    ..id = ''
    ..score = 0.0
    ..attempt = 0
    ..enteredScore = 0.0
    ..graderId = ''
    ..userId = ''
    ..submissionHistory = ListBuilder<Submission>()
    ..isGradeMatchesCurrentSubmission = false
    ..isLate = false
    ..excused = false
    ..missing = false;
}

File: libs/canvas-api-2/src/main/java/com/instructure/canvasapi2/models/Submission.kt
/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 *     Licensed under the Apache License, Version 2.0 (the "License");
 *     you may not use this file except in compliance with the License.
 *     You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *     Unless required by applicable law or agreed to in writing, software
 *     distributed under the License is distributed on an "AS IS" BASIS,
 *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *     See the License for the specific language governing permissions and
 *     limitations under the License.
 *
 */

package com.instructure.canvasapi2.models

import com.google.gson.annotations.SerializedName
import kotlinx.parcelize.Parcelize
import java.util.*

@JvmSuppressWildcards
@Parcelize
data class Submission(
        override var id: Long = 0,
        val grade: String? = null,
        val score: Double = 0.0,
        val attempt: Long = 0,
        @SerializedName("submitted_at")
        var submittedAt: Date? = null,
        @SerializedName("submission_comments")
        var submissionComments: List<SubmissionComment> = ArrayList(),
        val commentCreated: Date? = null,
        val mediaContentType: String? = null,
        val mediaCommentUrl: String? = null,
        val mediaCommentDisplay: String? = null,
        @SerializedName("submission_history")
        val submissionHistory: List<Submission?> = ArrayList(),
        val attachments: ArrayList<Attachment> = arrayListOf(),
        val body: String? = null,
        @SerializedName("rubric_assessment")
        var rubricAssessment: HashMap<String, RubricCriterionAssessment> = hashMapOf(),
        @SerializedName("grade_matches_current_submission")
        val isGradeMatchesCurrentSubmission: Boolean = false,
        @SerializedName("workflow_state")
        val workflowState: String? = null,
        @SerializedName("submission_type")
        val submissionType: String? = null,
        @SerializedName("preview_url")
        var previewUrl: String? = null,
        val url: String? = null,
        @SerializedName("late")
        val late: Boolean = false,
        @SerializedName("excused")
        val excused: Boolean = false,
        val missing: Boolean = false,
        @SerializedName("media_comment")
        val mediaComment: MediaComment? = null,
        // Conversation Stuff
        @SerializedName("assignment_id")
        val assignmentId: Long = 0,
        var assignment: Assignment? = null,
        @SerializedName("user_id")
        val userId: Long = 0,
        @SerializedName("grader_id")
        val graderId: Long = 0,
        val user: User? = null,
        // This value could be null. Currently will only be returned when getting the submission for
        // a user when the submission_type is discussion_topic
        @SerializedName("discussion_entries")
        val discussionEntries: ArrayList<DiscussionEntry> = arrayListOf(),
        // Group Info only available when including groups in the Submissions#index endpoint
        val group: Group? = null,
        @SerializedName("points_deducted")
        val pointsDeducted: Double? = null,
        @SerializedName("entered_score")
        val enteredScore: Double = 0.0,
        @SerializedName("entered_grade")
        val enteredGrade: String? = null,
        @SerializedName("posted_at")
        val postedAt: Date? = null,
        @SerializedName("grading_period_id")
        val gradingPeriodId: Long? = null
) : CanvasModel<Submission>() {
    override val comparisonDate get() = submittedAt
    override val comparisonString get() = submissionType

    val isWithoutGradedSubmission: Boolean get() = !isGraded && submissionType == null
    val isGraded: Boolean get() = grade != null

    /* Submissions will have dummy submissions if they grade an assignment with no actual submissions. We want to see if any are not dummy submissions */
    fun hasRealSubmission() = submissionHistory.any { it?.submissionType != null }
}
